src/express.re

```diff
@@ -1,43 +1,159 @@
-  should specify which JS type should be used.
-
-  An example of how to handle JS fns that take
-  more than one type, where some types are
-  containers and some primitives is the the
-  Response.send binding below.
-
-  sendString | sendJson | sendBuffer | sendArray
-  are all bindings to the .send method based on the
-  input type
-
-  tracked in https://github.com/BuckleTypes/bs-express/issues/6
- */
-module Next = {
+type done_;
+/** abstract type which ensure middleware function must either
+    call the [next] function or one of the [send] function on the 
+    response object. 
+
+    This should be a great argument for OCaml, the type system 
+    prevents silly error which in this case would make the server hang */ 
+
+/** TODO : maybe this should be a more common module like bs-error */
+module Error = {
+
   type t;
+  /** Error type */
+
+  external message : t => option string = "" 
+    [@@bs.send][@@bs.return null_undefined_to_opt];
+  external name : t => option string = "" 
+    [@@bs.send][@@bs.return null_undefined_to_opt];
 };
 
 module Request = {
   type t;
+  
+  type params = Js_dict.t Js_json.t;
+
+  external params : t => params = "" [@@bs.get];
+  /** [params request] return the JSON object filled with the 
+      request parameters */
+
+  external asJsonObject : t => Js_dict.t Js_json.t = "%identity";
+  /** [asJsonObject request] casts a [request] to a JSON object. It is 
+      common in Express application to use the Request object as a 
+      placeholder to maintain state through the various middleware which 
+      are executed. */ 
+
+  /** more property method coming later */
 };
 
 module Response = {
   type t;
-  external sendFile : t => string => 'a => unit = "" [@@bs.send];
-  external sendString : t => string => unit = "send" [@@bs.send];
-  external sendJson : t => Js.Json.t => unit = "send" [@@bs.send];
-  external sendBuffer : t => Buffer.t => unit = "send" [@@bs.send];
-  external sendArray : t => array 'a => unit = "send" [@@bs.send];
-  external json : t => Js.Json.t => unit = "" [@@bs.send] [@@ocaml.deprecated "Use sendJson instead`"];
+  external sendFile : t => string => 'a => done_ = "" [@@bs.send];
+  external sendString : t => string => done_ = "send" [@@bs.send];
+  external sendJson : t => Js.Json.t => done_ = "send" [@@bs.send];
+  external sendBuffer : t => Buffer.t => done_ = "send" [@@bs.send];
+  external sendArray : t => array 'a => done_ = "send" [@@bs.send];
+  external json : t => Js.Json.t => done_ = "" 
+    [@@bs.send] [@@ocaml.deprecated "Use sendJson instead`"];
 };
 
-module Express = {
+module Next : {
+  type content;
+
+  type t = (Js.undefined content => done_) [@bs];
+
+  let undefined : Js.undefined content;
+  /** value to use as [next] callback argument to invoke the next 
+      middleware */
+
+  let router : Js.undefined content;  
+  /** value to use as [next] callback argument to skip middleware 
+      processing. */ 
```

Comment:
This indicates (to me) that it will skip all middleware processing. Perhaps "... skip middleware processing for the current route" would be better?